native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
+       a single timestep. Remember there's overlap in the features, each timestep
+       contains N_CONTEXT past feature frames, the current feature frame, and
+       N_CONTEXT future feature frames, for a total of MFCC_WIN_LEN feature frames
+       per timestep.
+
+     - batch_buffer, used to buffer timesteps until there's enough data to compute
+       a batch of N_STEPS_PER_BATCH.
+
+     Data flows through all three buffers as audio samples are fed via the public
+     API. When audio_buffer is full, features are computed from it and pushed to
+     mfcc_buffer. When mfcc_buffer is full, the timestep is copied to batch_buffer.
+     When batch_buffer is full, we do a single step through the acoustic model
+     and accumulate results in StreamingState::accumulated_logits.
+
+     When fininshStream() is called, we decode the accumulated logits and return
+     the corresponding transcription.
+  */
+  StreamingState* setupStream(unsigned int prealloc_frames, unsigned int sample_rate);
+  void feedAudioContent(StreamingState* ctx, const short* buffer, unsigned int buffer_size);
+  char* finishStream(StreamingState* ctx);
+
+private:
+  /**
+   * @brief Perform decoding of the logits, using basic CTC decoder or
+   *        CTC decoder with KenLM enabled
+   *
+   * @param n_frames       Number of timesteps to deal with
+   * @param logits         Matrix of logits, of dimensions:
+   *                       [n_frames][batch_size][num_classes]
+   *
+   * @return String representing the decoded text.
+   *
+   * @note This method will free @p logits.
+   */
+  char* decode(std::vector<float>& logits);
+
+  /**
+   * @brief Do a single inference step in the acoustic model, with:
+   *          input=mfcc
+   *          input_lengths=[n_frames]
+   *
+   * @param mfcc batch input data
+   * @param n_frames number of timesteps in the data
+   *
+   * @param[out] output_logits Should be large enough to fit
+   *                           aNFrames * alphabet_size floats.
+   */
+  void infer(float* mfcc, int n_frames, std::vector<float>& output_logits);
+
+  void processAudioWindow(StreamingState* ctx, short* buf, unsigned int len);
+  void processMfccWindow(StreamingState* ctx, float* buf, unsigned int len);
+  void addZeroMfccWindow(StreamingState* ctx);
+  void processBatch(StreamingState* ctx, float* buf, unsigned int len);
 };
 
+StreamingState*
+Private::setupStream(unsigned int prealloc_frames,
+                     unsigned int /*sample_rate*/)
+{
+  Status status = session->Run({}, {}, {"initialize_state"}, nullptr);
+
+  if (!status.ok()) {
+    std::cerr << "Error running session: " << status << "\n";
+    return nullptr;
+  }
+
+  StreamingState* ctx = new StreamingState;
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+  ctx->accumulated_logits.reserve(prealloc_frames * BATCH_SIZE * num_classes);
+
+  memset(ctx->audio_buffer, 0, sizeof(ctx->audio_buffer));
```

Comment:
Since all those buffers are only internal, wouldn't it be simpler to use high-level data structure to manage them?

Reply:
I don't think it would make it any simpler for the audio buffer specifically, as we need to push data element by element to it. For the batch and MFCC buffers we could use std::vector, but it would still involve the same type of logic, copying fixed size inputs into an offset inside the buffers.